Controllers/Api/Dashboard/PowerbiReferenceController.cs

```diff
@@ -206,5 +289,271 @@ public async Task<IActionResult> Delete(int id)
+				if (powerbiReference == null)
+				{
+					var log = await logger.LogInvalidAccess(activity: activity, relevantObject: nameof(PowerbiReference), reference: $"{powerbiId}", log: $"Not found");
+					return this.BadRequest(log);
+				}
+
+				if (powerbiReference.Sharing?.UserShares == null)
+				{
+					var log = await logger.LogFailure(activity: activity, relevantObject: nameof(Sharing), reference: $"{powerbiId}", log: $"Not found");
+					return this.BadRequest(log);
+				}
+
+				powerbiReference.Sharing.UserShares = powerbiReference.Sharing.UserShares.Where(u => !userShares.Any(s => u.Email == s.Email)).ToList();
+
+				this.dataRepository.UpdatePowerbiReference(powerbiReference);
+
+				var result = await this.dataRepository.SaveChangesAsync();
+				if (!result)
+				{
+					var log = await logger.LogInvalidAccess(activity: activity, relevantObject: nameof(PowerbiReference), reference: $"{powerbiReference.Id}", log: $"Share failed to {string.Join(',', userShares.Select(s => s.Email))}");
+					return this.BadRequest(log);
+				}
+
+				await logger.LogDataAccess(activity: activity, relevantObject: nameof(PowerbiReference), reference: $"{powerbiReference.Id}", log: $"Share success to {string.Join(',', userShares.Select(s => s.Email))}");
+				return this.Ok(powerbiReference);
+			}
+			catch (Exception ex)
+			{
+				var log = await logger.LogException(activity, ex);
+				return this.BadRequest(log);
+			}
+
+			var logUnknown = await logger.LogUnknown(activity);
+			return this.BadRequest(logUnknown);
+		}
+
+		[HttpPut("group/share/{powerbiId}")]
+		public async Task<IActionResult> GroupShare([FromBody] List<GroupShare> shareTos, int powerbiId)
+		{
+			var activity = new ApplicationActivity() { Controller = ControllerName, Action = $"GroupShare/{powerbiId}" };
+
+			var owner = this.User.Identity.Name;
+
+			try
+			{
+				var hasPermission = await this.accessProfileHelper.HasPermission(owner, FeatureNames.Dashboard, Security.AccessLevelFlag.Write);
+				if (!hasPermission)
+				{
+					var log = await logger.LogInvalidAccess(activity: activity, relevantObject: nameof(PowerbiReference), log: $"Insufficient access");
+					return this.BadRequest("Insufficient access");
+				}
+
+				#region ServerSide Validation
+				if (!this.ModelState.IsValid)
+				{
+					var log = await logger.LogInvalidData(activity, "Invalid model");
+					return this.BadRequest(log);
+				}
+				#endregion ServerSide Validation
+
+				var powerbiReference = await this.dataRepository.GetPowerbiReferenceById(powerbiId);
+				if (powerbiReference == null)
+				{
+					var log = await logger.LogFailure(activity: activity, relevantObject: nameof(PowerbiReference), reference: $"{powerbiId}", log: $"Not found");
+					return this.BadRequest(log);
+				}
+
+				if (powerbiReference.Sharing?.GroupShares == null)
+				{
+					powerbiReference.Sharing = new Sharing()
+					{
+						GroupShares = shareTos
+					};
+				}
```

Comment:
This condition will never be true if there is Sharing object since GroupShares is initialized to be empty list
```suggestion
				if (powerbiReference.Sharing == null && !powerbiReference.Sharing.GroupShares.Any())
				{
					powerbiReference.Sharing = new Sharing()
					{
						GroupShares = shareTos
					};
				}
```